processing/src/main/java/org/apache/druid/query/scan/ScanQueryRunnerFactory.java

```diff
@@ -68,34 +89,214 @@ public ScanQueryRunnerFactory(
+            // (5) Join all the results into a single sequence
+
+            return Sequences.concat(
+                Sequences.map(
+                    Sequences.simple(groupedRunners),
+                    runnerGroup ->
+                        Sequences.map(
+                            Sequences.simple(runnerGroup),
+                            (input) -> Sequences.concat(
+                                Sequences.map(
+                                    input.run(queryPlus, responseContext),
+                                    srv -> Sequences.simple(srv.toSingleEventScanResultValues())
+                                )
+                            )
+                        ).flatMerge(
+                            seq -> seq,
+                            Ordering.from(new ScanResultValueTimestampComparator(
+                                query
+                            )).reverse()
+                        )
+                )
+            ).limit(
+                Math.toIntExact(query.getLimit())
+            );
+          }
+          throw new UOE(
+              "Time ordering for queries of %,d partitions per segment and a row limit of %,d is not supported."
+              + "  Try reducing the scope of the query to scan fewer partitions than the configurable limit of"
+              + " %,d partitions or lower the row limit below %,d.",
+              maxNumPartitionsInSegment,
+              query.getLimit(),
+              scanQueryConfig.getMaxSegmentPartitionsOrderedInMemory(),
+              scanQueryConfig.getMaxRowsQueuedForOrdering()
+          );
+        }
       }
     };
   }
 
+  @VisibleForTesting
+  Sequence<ScanResultValue> sortAndLimitScanResultValuesPriorityQueue(
+      Sequence<ScanResultValue> inputSequence,
+      ScanQuery scanQuery,
+      List<SegmentDescriptor> descriptorsOrdered
+  )
+  {
+    Comparator<ScanResultValue> priorityQComparator = new ScanResultValueTimestampComparator(scanQuery);
+
+    // Converting the limit from long to int could theoretically throw an ArithmeticException but this branch
+    // only runs if limit < MAX_LIMIT_FOR_IN_MEMORY_TIME_ORDERING (which should be < Integer.MAX_VALUE)
+    int limit = Math.toIntExact(scanQuery.getLimit());
+
+    PriorityQueue<ScanResultValue> q = new PriorityQueue<>(limit, priorityQComparator);
+
+    Yielder<ScanResultValue> yielder = inputSequence.toYielder(
+        null,
+        new YieldingAccumulator<ScanResultValue, ScanResultValue>()
+        {
+          @Override
+          public ScanResultValue accumulate(ScanResultValue accumulated, ScanResultValue in)
+          {
+            yield();
+            return in;
+          }
+        }
+    );
+    boolean doneScanning = yielder.isDone();
+    // We need to scan limit elements and anything else in the last segment
+    int numRowsScanned = 0;
+    Interval finalInterval = null;
+    while (!doneScanning) {
+      ScanResultValue next = yielder.get();
+      numRowsScanned++;
```

Comment:
`numRowsScanned++` should be in the `for (ScanResultValue srv : singleEventScanResultValues) {` loop below